[uss_qualifier] NET0460 evaluate display provider's flight data details endpoint performance

Shastick opened the pull request:
Note that validating the performance of the DP for returning details requires us to actually generate some queries for that endpoint.

To this end we query details for each flight that is returned when checking the `display_data` endpoint.

mickmis commented on monitoring/uss_qualifier/scenarios/astm/netrid/common/aggregate_checks.py:
```diff
@@ -71,9 +71,59 @@ def run(self):
 
+    def _dp_display_data_details_times_step(self):
+        """
+        Check performance of /display_data/<flight_id> requests and comfirm they conform to NetDpDetailsResponse95thPercentile (2s) and NetDpDetailsResponse99thPercentile (6s)
```
```suggestion
        Check performance of /display_data/<flight_id> requests and confirm they conform to NetDpDetailsResponse95thPercentile (2s) and NetDpDetailsResponse99thPercentile (6s)
```

mickmis commented on monitoring/uss_qualifier/scenarios/astm/netrid/common/aggregate_checks.py:
```diff
@@ -71,9 +71,59 @@ def run(self):
+                    )
+
+            self.record_note(
+                f"{participant}/display_data details stats computed on {len(durations)} queries",
```
```suggestion
                f"{participant}/display_data/<flight_id> stats computed on {len(durations)} queries",
```

mickmis commented on monitoring/uss_qualifier/scenarios/astm/netrid/display_data_evaluator.py:
```diff
@@ -261,6 +261,12 @@ def evaluate_system_instantaneously(
                     query,
                     verified_sps,
                 )
+                # We also issue queries to the flight details endpoint in order to collect
+                # performance statistics, which are computed and checked at a later stage.
+                if query.status_code == 200:
+                    for flight in observation.flights:
+                        (_, detailQuery) = observer.observe_flight_details(flight.id)
+                        self._test_scenario.record_query(detailQuery)
```
I'm a bit concerned about this addition, the polling in `evaluate_system_instantaneously` is invoked something like every second or so, and while that's not the case currently in the CI, nothing prevents a system from returning potentially tens of flights. Especially the fact that this is done sequentially is problematic, e.g. for v22a DPs could answer in 2 seconds and be within the limits. 
IMO we'd need a different approach, either:
- optimize this code (run concurrently, cache the results, etc.);
- implement it as a dedicated scenario where queries are made in a controlled manner;
- embed fetching details in an existing scenario where we can have more control over the queries being made
- or something else?

Shastick replied:
Questions on my mind:
 * An easy way to avoid too many queries is to grab the details of the first returned flight only: would that be acceptable?
 * Is it a problem if the polling is slowed down a bit in these situations? From my understanding we do the polling to wait until a certain state is reached. Is there anything else there?
 
By what I got from @BenjaminPelletier during one of our last weeklies, we'd rather keep things simple if possible for delivering the MVP and revisit if needed?

Shastick replied:
@mickmis we quickly broached the topic with @BenjaminPelletier: doing a single query is acceptable, and slowing the polling down slightly if the service is slow to respond is acceptable

Shastick commented on monitoring/uss_qualifier/resources/netrid/observers.py:
```diff
@@ -59,7 +59,10 @@ def observe_flight_details(
+            self.session,
+            "GET",
+            f"/display_data/{flight_id}",
+            scope=self.rid_version.read_scope,
```
The scope needs to be provided otherwise the request fails because of a failed authentication.

I suspect it was forgotten during a previous migration (other functions in the package have the scope passed)

BenjaminPelletier commented on monitoring/uss_qualifier/resources/netrid/observers.py:
```diff
@@ -59,7 +59,10 @@ def observe_flight_details(
+            self.session,
+            "GET",
+            f"/display_data/{flight_id}",
+            scope=self.rid_version.read_scope,
```
This is a call to the [InterUSS automated testing observation API](https://github.com/interuss/automated_testing_interfaces/blob/main/rid/v1/observation.yaml) and so we actually want [`dss.read.identification_service_areas`](https://github.com/interuss/automated_testing_interfaces/blob/e6695316e38607343cd16ba31175eac21a1e46b3/rid/v1/observation.yaml#L65) regardless of ASTM RID version even though that's not a great scope name for this endpoint.  Ideally, we would probably add this string as a scope constant (similar to [this](https://github.com/interuss/uas_standards/blob/79eba6dade57858122d998e9bf3e50e7e6cf5bb5/src/uas_standards/astm/f3411/v22a/constants.py#L4)) in [uas_standards](https://github.com/interuss/uas_standards/tree/main/src/uas_standards/interuss/automated_testing/rid/v1), but I think just a magic string constant is probably fine for now.

Shastick replied:
Adding the magic string to `rid.py`'s `RIDVersion` for now.

BenjaminPelletier replied:
This scope shouldn't be in RIDVersion -- RIDVersion differentiates between F3411-19 and F3411-22a.  This scope is not for either of those (even though its value happens, unfortunately, to match a scope defined in F3411-19) -- instead, it is the scope required to access the InterUSS RID observation automated testing API.  So, we can put a constant somewhere for now (or leave a magic string + TODO), but not RIDVersion :)  After [uas_standards is rev'd](https://github.com/interuss/uas_standards/pull/11), it will be `uas_standards.interuss.automated_testing.rid.v1.constants.Scope.Observe`.

Shastick replied:
Ah, my bad, I did a search and landed there so thought it was the same. Will update.

BenjaminPelletier commented on monitoring/uss_qualifier/scenarios/astm/netrid/v22a/aggregate_checks.md:
```diff
@@ -19,6 +19,14 @@ The observers to evaluate in the report.
+
+#### Performance of /display_data/<flight_id> requests check
+
+**[astm.f3411.v19.NET0460](../../../../requirements/astm/f3411/v19.md) Checks that the DP response times for the
+`/display_data/<flight_id>` endpoint have a p95 and p99 that are respectively below
```
nit: I would phrase the check more in terms of the requirement (Display Application flight details request) rather than how we happen to check it (`/display_data/<flight_id>` endpoint of InterUSS RID observation automated testing API):

```suggestion
**[astm.f3411.v19.NET0460](../../../../requirements/astm/f3411/v19.md) Checks that the DP response times for the
Display Application's flight details requests have a p95 and p99 that are respectively below
```

We're looking to justify why a failed check should necessarily indicate a failed requirement in this documentation.  Here's the text of the requirement (probably not optimally-written since area is ~irrelevant):

```
A Net-RID Display Provider shall (NET0460) respond
to requests from a Display Application for flight details for a
specific UAS within an area with a diagonal equal to or less
than [NetDetailsMaxDisplayAreaDiagonal] in
[NetDpDetailsResponse95thPercentile] seconds 95 % of the
time and in [NetDpDetailsResponse99thPercentile] seconds
99 % of the time.
```

BenjaminPelletier commented on monitoring/uss_qualifier/scenarios/astm/netrid/v22a/aggregate_checks.md:
```diff
@@ -19,6 +19,14 @@ The observers to evaluate in the report.
+
+#### Performance of /display_data/<flight_id> requests check
+
+**[astm.f3411.v19.NET0460](../../../../requirements/astm/f3411/v19.md) Checks that the DP response times for the
```
This is a v19 requirement, but it's listed in the v22a documentation.

Shastick commented:
Moving this to draft, as it will need some updates once https://github.com/interuss/monitoring/pull/159 is ready

Shastick commented:
Ready for another round of reviews (just updated on the latest master).

mickmis reviewed (approved):
Some minor issues to address before merge and some nits, otherwise LGTM.

Shastick commented on monitoring/uss_qualifier/resources/netrid/observers.py:
```diff
@@ -59,7 +59,10 @@ def observe_flight_details(
+            self.session,
+            "GET",
+            f"/display_data/{flight_id}",
+            scope=self.rid_version.dss_read_isa_scope,
```
Indeed, I was pondering that: I was just wondering if the scopes _should_ actually be different?

For v19 the read scope actually is the same as here (`dss.read.identification_service_areas`)

mickmis replied:
Nope they should not be different, they both refer to the same definition: https://github.com/interuss/automated_testing_interfaces/blob/main/rid/v1/observation.yaml#L182

BenjaminPelletier replied:
Agreed: both calls to /display_data and /display_data/{flight_id} use the same scope (which happens to be equal in value to an F3411-19 scope, but is semantically an InterUSS RID observation automated testing API scope)

Shastick commented on monitoring/uss_qualifier/scenarios/astm/netrid/v22a/aggregate_checks.md:
```diff
@@ -18,8 +18,16 @@ The observers to evaluate in the report.
+Display Application's flight details requests have a p95 and p99 that are respectively below
+`NetDpDetailsResponse95thPercentileSeconds` (2 seconds) and `NetDpDetailsResponse99thPercentileSeconds` (6 seconds).
+
 In this step, all successful display data queries made during the execution of the previous scenarios are aggregated per
+### Performance of /display_data requests test step
```
Dunno, moved that information into the test step description, it makes more sense there.

BenjaminPelletier commented on monitoring/monitorlib/rid.py:
```diff
@@ -65,6 +65,11 @@ def read_scope(self) -> str:
 
+    @property
+    def dss_read_isa_scope(self) -> str:
+        # Same regardless of ASTM RID version
```
[FYI](https://github.com/interuss/uas_standards/pull/11) (not a prereq for merging this PR)

Shastick commented:
Cleaned up `rid.py`, added a magic string with the correct substitution in a TODO for after https://github.com/interuss/uas_standards/pull/11 becomes available.

mickmis commented on monitoring/uss_qualifier/scenarios/astm/netrid/v22a/aggregate_checks.md:
```diff
@@ -18,8 +18,17 @@ The observers to evaluate in the report.
+`NetDpDetailsResponse95thPercentileSeconds` (2 seconds) and `NetDpDetailsResponse99thPercentileSeconds` (6 seconds).
+
 ### Performance of /display_data requests test step
-In this step, all successful display data queries made during the execution of the previous scenarios are aggregated per
 observer and per request (identified by their URLs). For each of those, and using the session length
```
A line of text went missing here

Shastick replied:
Thanks for pointing it out, looks like I need to improve my copy-pasta skills >.>

mickmis commented on monitoring/uss_qualifier/scenarios/astm/netrid/display_data_evaluator.py:
```diff
@@ -261,6 +261,21 @@ def evaluate_system_instantaneously(
+                            observation.flights[0].id
+                        )
+                        # Record query metadata for alter use in the aggregate checks
+                        detailQuery.server_id = observer.participant_id
+                        detailQuery.query_type = QueryType.flight_details(
+                            self._rid_version
+                        )
```
Well, that is a good point. It does indeed seem to be hardcoded to v19.

I think that to get to something consistent we should either:
1. get rid altogether of  `RIDSystemObserver.rid_version` and as you propose pass `rid_version` as a function parameter
(in which case BTW the call `min_dim = 2 * observer.rid_version.min_obfuscation_distance_m` in `display_data_evaluator.py` must be updated
2. or properly provision `RIDSystemObserver.rid_version` with a parameter taken from the yaml configuration

IMO 2. would be slightly preferable but it requires some additional refactoring (and updates of yaml configs), but 1. is alright as well since this `RIDSystemObserver.rid_version` does not seem to be in use, as  `RIDObservationEvaluator._rid_version` is used basically everywhere.

BenjaminPelletier commented:
@mickmis : feel free to merge when ready & you think is appropriate
